File: LaddersAndSlides_GameEngine/Domain/Player.cs
namespace LaddersAndSlides_GameEngine.Domain
{
    using System;

    public class Player
    {
        public string Name { get; set; }
        public Uri ImageUri { get; set; }
        public int CurrentTileNumber { get; set; }
        public int CurrentMovesRemaining { get; set; }
        public int TurnOrder { get; set; }
        public bool TurnInProcess { get; set; }
        public bool CurrentlyOnAlternateRow { get; set; }

        public override bool Equals(object obj)
        {
            if (ReferenceEquals(null, obj)) return false;
            if (ReferenceEquals(this, obj)) return true;
            if (obj.GetType() != GetType()) return false;
            return Equals((Player) obj);
        }

        protected bool Equals(Player other)
        {
            return string.Equals(Name, other.Name) && TurnOrder == other.TurnOrder;
        }

        public override int GetHashCode()
        {
            unchecked
            {
                return (Name.GetHashCode() * 397) ^ TurnOrder;
            }
        }

        public static bool operator ==(Player left, Player right)
        {
            return Equals(left, right);
        }

        public static bool operator !=(Player left, Player right)
        {
            return !Equals(left, right);
        }
    }
}

File: src/domain/Player.ts
import { PlayerColor } from './PlayerColor';

export class Player {
  name: string = '';
  imageUri: string = '';
  currentTileNumber: number = 0;
  currentGridRow: number = 0;
  currentGridColumn: number = 0;
  currentMovesRemaining: number = 0;
  turnOrder: number = 0;
  turnInProcess: boolean = false;
  currentlyOnAlternateRow: boolean = false;
  specialMoveTransportDestination: number = 0;
  gameTokenColor: PlayerColor = PlayerColor.Blue;

  constructor(init?: Partial<Player>) {
    if (init) {
      Object.assign(this, init);
    }
  }

  equals(other: Player | null): boolean {
    if (!other) return false;
    if (this === other) return true;
    return this.name === other.name && this.turnOrder === other.turnOrder;
  }

  static equals(left: Player | null, right: Player | null): boolean {
    if (left === null && right === null) return true;
    if (left === null || right === null) return false;
    return left.equals(right);
  }
}
